# BFS

- Modeling
- Find Neighbors
- Time complexity of Dijktra: O($E*logV$)

## 993. Cousins in Binary Tree
[993. Cousins in Binary Tree](https://leetcode.com/problems/cousins-in-binary-tree/)

In [47]:
from collections import deque
q = deque()
q.append(5)
q.append(6)
print(q)

# doubly linked list: check the existence need O(n)
if 88 in q:
    print("yes")

deque([5, 6])


In [48]:
def isCousins(self, root, x: int, y: int) -> bool:
    if not root:
        return False
    node = (root, float('inf'))
    q = deque([node])
    while q:
        size = len(q)
        par = []
        # add all siblings and cousions in q
        for _ in range(size):
            cur, parent = q.popleft()
            if cur.val == x or cur.val == y:
                par.append(parent)
            if cur.left:
                q.append((cur.left, cur.val))
            if cur.right:
                q.append((cur.right, cur.val))
        # check the parent list
        if len(par) == 2:
             if par[0] != par[1]:
                return True
             return False

## 785. Is Graph Bipartite?
[785. Is Graph Bipartite?](https://leetcode.com/problems/is-graph-bipartite/)

In [4]:
res = [[0 for _ in range(3)] for _ in range(4)]
res

[[0, 0, 0], [0, 0, 0], [0, 0, 0], [0, 0, 0]]

In [3]:
def isBipartite(self, graph) -> bool:
    """
    BFS: BFS for all nodes (not connected case)
    visited = [0 for _ in range(n)]
    """
    if not graph:
        return True
    n = len(graph)
    visited = [0 for _ in range(n)]
    for i in range(n):
        if len(graph[i]) == 0 or visited[i] != 0:
            continue
        if not self.bfs(i, graph, visited):
            return False
    return True

def bfs(self, cur, graph, visited):
    q = deque([cur])
    visited[cur] = 1
    while q:
        node = q.popleft()
        # generate neighbors
        for item in graph[node]:
            if visited[item] == 0:
                visited[item] = -1 * visited[node]
                q.append(item)
            if visited[item] != 0 and visited[item] == visited[node]:
                return False
    return True

## 200. Number of Islands
[200. Number of Islands](https://leetcode.com/problems/number-of-islands/)

In [60]:
# BFS
class Solution:
    def numIslands(self, grid) -> int:
        count = 0
        m = len(grid)
        n = len(grid[0])
        directions = [[0, 1],[1, 0],[0,-1],[-1, 0]]
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    grid[i][j] = '0'
                    self.bfs(grid, i, j, directions)
                    count += 1
        return count
    
    def bfs(self, grid, i, j, directions):
        q = deque([[i,j]])
        while q:
            row, col = q.popleft()
            for dr, dc in directions:
                r = row + dr
                c = col + dc
                if r >= 0 and r < len(grid) and c >= 0 and c < len(grid[0]) and grid[r][c] == '1':
                    grid[r][c] = '0'
                    q.append([r, c])

## 1197. Minimum Knight Moves
[1197. Minimum Knight Moves](https://leetcode.com/problems/minimum-knight-moves/)

## 994. Rotting Oranges
[994. Rotting Oranges](https://leetcode.com/problems/rotting-oranges/)

In [40]:
class Solution:
    def orangesRotting(self, grid) -> int:
        """
        1. Pre-processing: calculate the fresh_num and put rotten num into q
        2. Edge case: no fresh return 0, no rotten return -1
        3. BFS
            generate fresh orange into q
            change grid -> 2: dedup in-place
            step += 1
        4. check fresh_num 
           return step - 1 !!!!
        """
        m = len(grid)
        n = len(grid[0])
        q = deque()
        fresh_num = 0
        # pre-processing: calculate the fresh_num and put rotten num into q
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    q.append([i, j])
                elif grid[i][j] == 1:
                    fresh_num += 1
        
        #edge case
        # no fresh orange
        if fresh_num == 0:
            return 0
        
        # no rotten orange
        if len(q) == 0:
            return -1
        
        # bfs
        step = 0
        directions = [[0, 1], [0, -1], [1, 0],[-1, 0]]
        while q:
            size = len(q)
            for _ in range(size):
                cur = q.popleft()
                row = cur[0]
                col = cur[1]
                
                # generate neighbors
                neighbors = self.findNeighbors(grid, m, n, row, col, directions)
                fresh_num -= len(neighbors)
                for r, c in neighbors:
                    q.append([r, c])
            step += 1
        
        if fresh_num > 0:
            return -1
        
        return step - 1
    
    def findNeighbors(self, grid, m, n, row, col, directions):
        neighbors = []
        for dr, dc in directions:
            r = row + dr
            c = col + dc
            if r >= 0 and r < m and c >= 0 and c < n and grid[r][c] == 1:
                grid[r][c] = 2 # do not need deduplication (visited)
                neighbors.append([r, c])
        return neighbors

## 317. Shortest Distance from All Buildings
[317. Shortest Distance from All Buildings](https://leetcode.com/problems/shortest-distance-from-all-buildings/)

In [42]:
a = [[2, 3, 4,5, 6],[7,3,4,2,1]]
res = float('inf')
for i in range(len(a)):
    res = min(min(a[i]), res)
res

1

In [44]:
a= [[1, 2],[3, 4]]
for i, j in a:
    print(i, j)

1 2
3 4


## 286. Walls and Gates
[286. Walls and Gates](https://leetcode.com/problems/walls-and-gates/)

In [52]:
q = deque()
q.append([1,2,3])
q.append([4,5,6])
a, b, c = q.popleft()
print(a, b, c)

1 2 3


In [56]:
class Solution:
    def wallsAndGates(self, rooms) -> None:
        """
        Do not return anything, modify rooms in-place instead.
        
        For each gate call bfs and matain the minimum val to the empty room
        """
        if not rooms:
            return
        m = len(rooms)
        n = len(rooms[0])
        directions = [[1, 0],[-1, 0],[0, 1],[0,-1]]
        q = deque()
        for i in range (m):
            for j in range (n):
                if rooms[i][j] == 0:
                    q.append([i, j])
        visited = [[False for _ in range(n)] for _ in range(m)]
        self.bfs(rooms, visited, q, directions)
    
    def bfs(self, rooms, visited, q, directions):
        while q:
            size = len(q)
            for i in range(size):
                row, col = q.popleft()
                for d in directions:
                    r = row + d[0]
                    c = col + d[1]
                    if r >= 0 and r < len(rooms) and c >= 0 and c < len(rooms[0]):
                        if not visited[r][c] and rooms[r][c] > 0:
                            visited[r][c] = True
                            q.append([r, c])
                            rooms[r][c] = rooms[row][col] + 1

## 417. Pacific Atlantic Water Flow
[417. Pacific Atlantic Water Flow](https://leetcode.com/problems/pacific-atlantic-water-flow/)

## 490. The Maze
[490. The Maze](https://leetcode.com/problems/the-maze/)

In [25]:
a = [[False for _ in range(3)] for _ in range(3)]
b = [[1,2],[3,4]]
for row, col in b:
    print(row, col)

1 2
3 4


In [31]:
class Solution:
    def hasPath(self, maze, start, destination) -> bool:
        """
        Build graph: 
            Node(State): the position that the ball stop (intermidiate position won't be the node)
            Edge: 4 directions
        
        """
        m = len(maze)
        n = len(maze[0])
        # initial state
        q = deque([(start[0], start[1])])
        visited =  [[False for _ in range(n)] for _ in range(m)]
        visited[start[0]][start[1]] = True
        directions = [[0, 1],[0, -1],[-1, 0],[1, 0]]
        while q:
            cur = q.popleft()
            row = cur[0]
            col = cur[1]
            
            # check destination
            if row == destination[0] and col == destination[1]:
                return True
            
            # get neighbors
            neighbors = self.getNeighbors(maze, row, col, directions)
            for r, c in neighbors:
                if not visited[r][c]:
                    visited[r][c] = True
                    q.append([r, c])
        return False
        
    def getNeighbors(self, maze, row, col, directions):
        neighbors = []
        for dr, dc in directions:
            r = row
            c = col
            # didn't go to this direction before
            # go striaight 
            while r + dr >= 0 and r + dr < len(maze) and c + dc >= 0 and c + dc < len(maze[0]) and maze[r + dr][c + dc] == 0:
                # not visited check here!!!!
                r += dr
                c += dc
            neighbors.append([r, c])
        return neighbors

# Implicit Graph

## 773. Sliding Puzzle
[773. Sliding Puzzle](https://leetcode.com/problems/sliding-puzzle/)

- Follw up: thousands of queries:
    - Start and End is interchangable (undirected graph)
    - End is fixed. Start = End, run BFS using a hashmap to record all the moves start to all the configurations.

In [5]:
class Solution:
    def slidingPuzzle(self, board) -> int:
        """
        1. Each configuration of the board represents a Node
        2, Edge: unweighted 
        3. Successor function: 2 or 3 directions to move 0
        
        Implementation:
        BFS
        Hashset: visited
        check if we get to the target Node tuple(1,2,3,4,5,0)
        keep track of the minimum moves
        
        # matrix --> list
          row = idx // c 
          col = idx % c
        """
        m = len(board)
        n = len(board[0])
        end = (1,2,3,4,5,0)
        directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]
        # Pre-processing input: convert matrix to string
        start = tuple(board[0] + board[1]) # !!! tuple is hashable!
        # print(start)
        visited = set([start])
        q = deque([start])
        moves = -1
        # BFS
        # neis = self.findNeighbors(m, n, start, directions)
        # print(neis)
        while q:
            size = len(q)
            moves += 1
            for i in range(size):
                cur = q.popleft()
                if cur == end:
                    return moves
                # find neighbors
                neighbors = self.findNeighbors(m, n, cur, directions)
                for nei in neighbors:
                    if nei not in visited:
                        q.append(nei)
                        visited.add(nei)
        return -1
    
    def findNeighbors(self, m, n, cur, directions):
        neighbors = []
        cur_list = list(cur)
        # print(cur_list)
        index_zero = cur_list.index(0)
        # print(index_zero)
        row = index_zero // n
        col = index_zero % n
        # print(row, col)
 
        # print(cur_list)
        for dr, dc in directions:
            r = row + dr
            c = col + dc
            if r >= 0 and r < m and c >= 0 and c < n:
                new_index = n * r + c
                cur_list[new_index], cur_list[index_zero] = cur_list[index_zero], cur_list[new_index]
                neighbors.append(tuple(cur_list))
                cur_list[new_index], cur_list[index_zero] = cur_list[index_zero], cur_list[new_index]
        return neighbors

## 752. Open the Lock
[752. Open the Lock](https://leetcode.com/problems/open-the-lock/)


- Complexity: $ O(N^2 * A^N + D)$



```
N is Number of dials (4 in our case)
A is number of alphabets (10 in our case -> 0 to 9)
D is the size of deadends.

There are 10 x 10 x 10 x 10 possible combinations => 10^4 => A^N
For each combination, we are looping 4 times (which is N) and in each iteration, there are substring operations ( which is O(N) * constant) => O(4N*constant) => O(4N) => O(NN) => O(N^2)
Total complexity => A^N * N^2, plus D to create the hashset => N^2 * A^N + D
```

In [ ]:
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        """
        State: a str represents a state
        Successor function: +1 / -1 for each digit (8)
        Use a hashset to avoid to go to the deadends
        Use the same hashset to deduplication
        each level -> step
        """
        # edge case
        if target == '0000':
            return 0
        # edd deadends to the hashset
        visited = set(deadends)
        # print(visited)
        if '0000' in visited:
            return -1
        # bfs
        q = deque(['0000'])
        visited.add('0000')
        steps = 0
        while q:
            size = len(q)
            for _ in range(size):
                cur = q.popleft()
                if cur == target:
                    return steps
                # add neighbors
                # neighbors = self.getNeighbors(cur)
                neighbors = self.successor(cur)
                for nei in neighbors:
                    if nei not in visited:
                        q.append(nei)
                        visited.add(nei)
            steps += 1
        return -1
    
    # Works
    def getNeighbors(self, cur):
        neis = []
        for i in range(len(cur)):
            chars = list(cur)
            digit = int(chars[i])
            # + 1
            a = -1 if digit == 9 else digit
            chars[i] = str(a + 1)
            adds = "".join(chars)
            neis.append(adds)
            
            # - 1
            b = 10 if digit == 0 else digit
            chars[i] = str(b - 1)
            minus = "".join(chars)
            neis.append(minus)
        return neis
    
    # Works
    def successor(self, state):  
        res = []  
        state_list = list(state) 
        # state_list = [int(x) for x in state_list] 

        for i in range(len(state_list)):  
            increase = chr(ord(state_list[i]) + 1) if state_list[i] != '9' else '0' 
            state_list[i] = increase 
            res.append(''.join(state_list)) 
            state_list[i] = state[i]   
            
            decrease =  chr(ord(state_list[i]) - 1) if state_list[i] != '0' else '9' 
            state_list[i] = decrease 
            res.append(''.join(state_list)) 
            state_list[i] = state[i]   

        return res

In [23]:
# Concatinate str
cur = '123'
# b = int(cur[1])
# adds = cur[:1] + str(b + 1) + cur[1 + 1:]
# minus = cur[:1] + str(b - 1) + cur[1 + 1:]
# print(adds, minus)

chars = list(cur)
print("list", chars)
ch = int(chars[1])
print("ch", ch)
chars[1] = str(ch + 1)
print("new_list", chars)
new_s = "".join(chars)
new_s

list ['1', '2', '3']
ch 2
new_list ['1', '3', '3']


'133'

In [11]:
digit = 8
digit = 10 if digit == 0 else digit
digit

8

# Topological Sort

## 207. Course Schedule
[207. Course Schedule](https://leetcode.com/problems/course-schedule/)

## 210. Course Schedule II
[210. Course Schedule II](https://leetcode.com/problems/course-schedule-ii/)

## 953. Verifying an Alien Dictionary
[953. Verifying an Alien Dictionary](https://leetcode.com/problems/verifying-an-alien-dictionary/)

## 269. Alien Dictionary
[269. Alien Dictionary](https://leetcode.com/problems/alien-dictionary/)

# Dijkstra

## 378. Kth Smallest Element in a Sorted Matrix
[378. Kth Smallest Element in a Sorted Matrix](https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/)

In [39]:
class Solution:
    def kthSmallest(self, matrix, k: int) -> int:
        """
        BFS: generate right and buttom elements
        when expanding a node, k -= 1
        when k == 0
        it's the result
        """
        m = len(matrix)
        n = len(matrix[0])
        q = [[matrix[0][0], 0, 0]]
        heapq.heapify(q)
        visited = [[False for _ in range(n)] for _ in range(m)]
        visited[0][0] = True
        directions = [[0, 1], [1, 0]]
        while q:
            cur = heapq.heappop(q)
            val = cur[0]
            r = cur[1]
            c = cur[2]
            k -= 1
            # check if we find the result
            if k == 0:
                return val
            # find neighbors
            neighbors = self.findNeighbors(matrix, m, n, r, c, directions)
            for row, col in neighbors:
                if not visited[row][col]:
                    heapq.heappush(q, [matrix[row][col], row, col])
                    visited[row][col] = True
        return -1
    
    def findNeighbors(self, matrix, m, n, row, col, directions):
        neighbors = []
        for dr, dc in directions:
            r = row + dr
            c = col + dc
            if r >= 0 and r < m and c >= 0 and c < n:
                neighbors.append([r, c])
        return neighbors

## 318. Maximum Product of Word Lengths
[318. Maximum Product of Word Lengths](https://leetcode.com/problems/maximum-product-of-word-lengths/)

## 2040. Kth Smallest Product of Two Sorted Arrays
[2040. Kth Smallest Product of Two Sorted Arrays](https://leetcode.com/problems/kth-smallest-product-of-two-sorted-arrays/)

## 505. The Maze II
[505. The Maze II](https://leetcode.com/problems/the-maze-ii/)

In [ ]:
# TC = O(log(n) * m * n)
# SC = O(m*n)
class Solution:
    def shortestDistance(self, maze, start, destination) -> int:
        """
        1. Graph:
            Nodes: the positions that the ball stops 
                   Use a tuple to maintain the cordinate and current distance(distance, row, col)
            Edge: 4 directions
            Weight: the weight of the node is the distance from the neighbor node
        
        2. Uniform Cost search from start to destination (Dijkstra)
        3. The same node might be generated many times if we find a shorter distance
        4. Deduplication: 
            visited[r][c] represent the current shortest distance
            if the distance doesn't change, we won't visited the node anymore
        """
        m = len(maze)
        n = len(maze[0])
        visited = [[float('inf') for _ in range(n)] for _ in range(m)]
        q = [(0, start[0], start[1])]
        visited[start[0]][start[1]] = 0
        heapq.heapify(q)
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        while q:
            cur = heapq.heappop(q)
            dis = cur[0]
            row = cur[1]
            col = cur[2]
            
            # check current node
            if row == destination[0] and col == destination[1]:
                return dis
            
            # generate neighbors
            neighbors = self.findNeighbors(maze, m, n, row, col, dis, directions)
            for r, c, dis in neighbors:
                if visited[r][c] > dis:
                    heapq.heappush(q, (dis, r, c))
                    visited[r][c] = dis
        return -1
        
    def findNeighbors(self, maze, m , n, row, col, dis, directions):
        neighbors = []
        for dr, dc in directions:
            r = row
            c = col
            distance = dis
            while (r + dr) >= 0 and (r + dr) < m and (c + dc) >= 0 and (c + dc) < n and maze[r + dr][c + dc] == 0:
                r += dr
                c += dc
                distance += 1
            neighbors.append((r, c, distance))
        return neighbors

## 499. The Maze III
[499. The Maze III](https://leetcode.com/problems/the-maze-iii/)

In [36]:
visited = [[[float('inf'),""] for _ in range(3)] for _ in range(3)]
visited

print("lu" < "lul")

True


In [37]:
class Solution:
    def findShortestWay(self, maze, ball, hole) -> str:
        """
        Same as Dijkstra solution in Maze 2
        Custom heapq -> distance > path
        Hole is a stop position
        """
        m = len(maze)
        n = len(maze[0])
        visited = [[[float('inf'),""] for _ in range(n)] for _ in range(m)]
        # initial state
        q = [Node(ball[0], ball[1], 0, "")]
        heapq.heapify(q)
        visited[ball[0]][ball[1]] = [0,""]
        directions = [[0, 1, "r"], [0, -1, "l"], [1, 0, "d"],[-1, 0, "u"]]
        res = []
        while q:
            cur = heapq.heappop(q)
            # print(cur.path, cur.distance)
            # check if get to the hole
            if cur.x == hole[0] and cur.y == hole[1]:
                return cur.path
            
            # generate neighbors
            neighbors = self.findNeighbors(maze, m, n, cur, hole, directions)
            for x, y, dis, path in neighbors:
                if visited[x][y][0] > dis or visited[x][y][0] == dis and visited[x][y][1] > path:
                    heapq.heappush(q, Node(x, y, dis, path))
                    visited[x][y][0] = dis
                    visited[x][y][1] = path
                    
        return "impossible"
    
    def findNeighbors(self, maze, m, n, cur, hole, directions):
        neighbors = []
        for dr, dc, p in directions:
            r = cur.x
            c = cur.y
            dis = cur.distance
            path = cur.path
            while (r + dr) >= 0 and (r + dr) < m and (c + dc) >= 0 and (c + dc) < n and maze[r+dr][c+dc] == 0:
                r += dr
                c += dc
                dis += 1
                # check if the ball passes by the hole
                if r == hole[0] and c == hole[1]:
                    neighbors.append((r, c, dis, path + p))
                    break
            neighbors.append((r, c, dis, path + p))
        return neighbors
    
class Node:
    def __init__(self, x, y, distance, path):
        self.x = x
        self.y = y
        self.distance = distance
        self.path = path
    
    def __lt__(self, node2):
        if self.distance == node2.distance:
            return self.path < node2.path
        return self.distance < node2.distance

In [38]:
class Solution:
    def findShortestWay(self, maze, ball, hole) -> str:
        """
        Same as Dijkstra solution in Maze 2
        Custom heapq -> distance > path
        Hole is a stop position
        """
        m = len(maze)
        n = len(maze[0])
        visited = [[float('inf') for _ in range(n)] for _ in range(m)]
        # initial state
        q = [Node(ball[0], ball[1], 0, "")]
        heapq.heapify(q)
        visited[ball[0]][ball[1]] = 0
        directions = [[0, 1, "r"], [0, -1, "l"], [1, 0, "d"],[-1, 0, "u"]]
        res = []
        while q:
            cur = heapq.heappop(q)
            # print(cur.path, cur.distance)
            # check if get to the hole
            if cur.x == hole[0] and cur.y == hole[1]:
                return cur.path
            
            # generate neighbors
            neighbors = self.findNeighbors(maze, m, n, cur, hole, directions)
            for x, y, dis, path in neighbors:
                if visited[x][y] >= dis:
                    heapq.heappush(q, Node(x, y, dis, path))
                    visited[x][y] = dis
                    
        return "impossible"
    
    def findNeighbors(self, maze, m, n, cur, hole, directions):
        neighbors = []
        for dr, dc, p in directions:
            r = cur.x
            c = cur.y
            dis = cur.distance
            path = cur.path
            while r >= 0 and r < m and c >= 0 and c < n and maze[r][c] == 0:
                r += dr
                c += dc
                dis += 1
                # check if the ball passes by the hole
                if r == hole[0] and c == hole[1]:
                    neighbors.append((r, c, dis, path + p))
                    break
            # check if we don't move (or !!!!!)
            if r - dr != cur.x or c - dc != cur.y:
                neighbors.append((r - dr, c - dc, dis - 1, path + p))
        return neighbors
    
class Node:
    def __init__(self, x, y, distance, path):
        self.x = x
        self.y = y
        self.distance = distance
        self.path = path
    
    def __lt__(self, node2):
        if self.distance == node2.distance:
            return self.path < node2.path
        return self.distance < node2.distance

## 787. Cheapest Flights Within K Stops
[787. Cheapest Flights Within K Stops](https://leetcode.com/problems/cheapest-flights-within-k-stops/)

In [4]:
class Solution:
    def findCheapestPrice(self, n: int, flights, src: int, dst: int, k: int) -> int:
        """
        Node: (dis, num, cur_k)
        Neighbors: (num, weight)
        Visited: dict((num, [cur_dis, cur_k]))
        """
        graph = self.buildGraph(flights)
        # print(graph)
        visited = defaultdict(list)
        visited[src] = [0, k + 1]
        pq = [(0, src, k + 1)]
        heapq.heapify(pq)
        while pq:
            cur = heapq.heappop(pq)
            cur_dis = cur[0]
            cur_src = cur[1]
            cur_k = cur[2]
            
            if dst == cur_src:
                return cur_dis
            
            if cur_k > 0:
                neighbors = graph[cur_src]
                # print("neis", neighbors)
                for num, weight in neighbors:
                    new_node = (cur_dis + weight, num, cur_k - 1)
                    if num not in visited:
                        visited[num] = [cur_dis + weight, cur_k - 1]
                        heapq.heappush(pq, new_node)
                    else:
                        prev_dis = visited[num][0]
                        prev_k = visited[num][1]
                        # if prev_dis > cur_dis + weight or prev_k < cur_k - 1:
                        if prev_dis != cur_dis + weight: #or prev_k < cur_k - 1:
                            visited[num] = [cur_dis + weight, cur_k - 1]
                            heapq.heappush(pq, new_node) 

        return -1

    def buildGraph(self, flights):
        # (num, [(des1, weight1), (des2, weight2)])
        graph = defaultdict(list)
        for cur in flights:
            graph[cur[0]].append((cur[1], cur[2]))
        return graph

In [3]:
# Time exceed
class Solution:
    def findCheapestPrice(self, n: int, flights, src: int, dst: int, k: int) -> int:
        """
        Node: (dis, num, cur_k)
        Neighbors: (num, weight)
        Visited: set((cur_dis, cur_src, cur_k))
        """
        graph = self.buildGraph(flights)
        # print(graph)
        visited = set()
        pq = [(0, src, k + 1)]
        heapq.heapify(pq)
        while pq:
            cur = heapq.heappop(pq)
            print("cur", cur)
            visited.add(cur)
            
            cur_dis = cur[0]
            cur_src = cur[1]
            cur_k = cur[2]
            
            if dst == cur_src:
                return cur_dis
            
            if cur_k == 0:
                continue

            if cur_k > 0:
                neighbors = graph[cur_src]
                # print("neis", neighbors)
                for num, weight in neighbors:
                    # print("nums", num)
                    # print("weight", weight)
                    new_node = (cur_dis + weight, num, cur_k - 1)
                    if new_node not in visited:
                        heapq.heappush(pq, new_node)

        return -1

    def buildGraph(self, flights):
        # (num, [(des1, weight1), (des2, weight2)])
        graph = defaultdict(list)
        for cur in flights:
            graph[cur[0]].append((cur[1], cur[2]))
        return graph

## 743. Network Delay Time
[743. Network Delay Time](https://leetcode.com/problems/network-delay-time/)

In [6]:
class Solution:
    def networkDelayTime(self, times, n: int, k: int) -> int:
        """
        Weighted Directed Graph
        BuildGraph -> adjacency list
        Node: (cur_time, num)
        Dijktra --> until all the nodes are visited
        Visited: dict(key = num, value: cur_time)
        check expand a node, add to a final check set() == n
        """
        
        # build graph
        graph = self.buildGraph(times)
        # print(graph)
        # intial state
        pq = [(0, k)]
        heapq.heapify(pq)
        visited = defaultdict(int)
        visited[k] = 0
        received = set()
        while pq:
            cur = heapq.heappop(pq)
            cur_time = cur[0]
            num = cur[1]
            received.add(num)
            
            #check the termination condition
            if len(received) == n:
                return cur_time
            
            # find neighbors
            neighbors = graph[num]
            for nei, weight in neighbors:
                if nei not in visited:
                    visited[nei] = cur_time + weight
                    heapq.heappush(pq, (cur_time + weight, nei))
                else:
                    if cur_time + weight < visited[nei]:
                        visited[nei] = cur_time + weight
                        heapq.heappush(pq, (cur_time + weight, nei))
        return -1
                
    
    def buildGraph(self, times):
        graph = defaultdict(list)
        for u, v, w in times:
            graph[u].append((v, w))
        return graph

## 127. Word Ladder
[127. Word Ladder](https://leetcode.com/problems/word-ladder/)

In [13]:
# chr() and ord()
a = ord('a')
print(a)
for i in range(1):
    print(chr(a))

97
a


In [ ]:
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        """
        Node: each word in the wordlist
        Edge: differ by a single word
        
        Build Graph: for each digit, modify to 26 letters(exclude itself and check if it exist in the wordlist
        
        BFS1
        """
        word_dict = set(wordList)
        # edge case
        if endWord not in word_dict:
            return 0
        # initial
        q = deque([beginWord])
        visited = set([beginWord])
        moves = 0
        # neis = self.findNeighbors("cog", word_dict)
        # print(neis)
        while q:
            size = len(q)
            moves += 1
            for _ in range(size):
                cur = q.popleft()
                if cur == endWord:
                    return moves
                neighbors = self.findNeighbors(cur, word_dict)
                for nei in neighbors:
                    if nei not in visited:
                        q.append(nei)
                        visited.add(nei)
        return 0
    
    def findNeighbors(self, word, word_dict):
        neighbors = []
        cur = list(word)
        for i in range(len(cur)):
            original = cur[i]
            # print(original)
            a = ord('a')
            for j in range(26):
                new_letter = chr(j + a)
                # print(new_letter)
                if new_letter != original:
                    cur[i] = new_letter
                    new_word = "".join(cur)
                    # print(new_word)
                    if new_word in word_dict:
                        neighbors.append(new_word)
            cur[i] = original
        return neighbors

## 126. Word Ladder II
[126. Word Ladder II](https://leetcode.com/problems/word-ladder-ii/)

## 79. Word Search
[79. Word Search](https://leetcode.com/problems/word-search/)

## 212. Word Search II
[212. Word Search II](https://leetcode.com/problems/word-ladder-ii/)

## 1293. Shortest Path in a Grid with Obstacles Elimination
[1293. Shortest Path in a Grid with Obstacles Elimination](https://leetcode.com/problems/shortest-path-in-a-grid-with-obstacles-elimination/)

In [1]:
class Solution:
    def shortestPath(self, grid, k: int) -> int:
        """
        Node (i, j, cur_k)
        Call BFS
        ！！！！ cur_k should be in the state
        we might visit the same cell but if the k is different, the path and distance are also different!
        """
        m = len(grid)
        n = len(grid[0])
        target = (m - 1, n - 1)
        q = deque([(0, 0, k, 0)])
        visited = set()
        visited.add((0,0,k))
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        while q:
            cur = q.popleft()
            row = cur[0]
            col = cur[1]
            cur_k = cur[2]
            dis = cur[3]
            
            if (row, col) == target:
                return dis
            
            neighbors = self.findNeighbors(grid, m, n, row, col, directions)
            for r, c in neighbors:
                k_eli = cur_k - grid[r][c]
                if k_eli >= 0:
                    if (r, c, k_eli) not in visited:
                        q.append((r, c, k_eli, dis + 1))
                        visited.add((r, c, k_eli))
        return -1
    
    def findNeighbors(self, grid, m, n, row, col, directions):
        neighbors = []
        for dr, dc in directions:
            r = row + dr
            c = col + dc
            if 0 <= r < m and 0 <= c < n:
                neighbors.append((r, c))
        return neighbors

## 1263. Minimum Moves to Move a Box to Their Target Location
[1263. Minimum Moves to Move a Box to Their Target Location](https://leetcode.com/problems/minimum-moves-to-move-a-box-to-their-target-location/)

## 489. Robot Room Cleaner
[489. Robot Room Cleaner](https://leetcode.com/problems/robot-room-cleaner/)

# Graph

- Unidirected Graph
- Directed Graph
- Weighted Graph
- Bipartite Graph
- Connected Graph
- DAG: directed acyslic Graph
- Circle and Cycle: circle(shape 圈）and Cycle(order 闭环）

## 261. Graph Valid Tree
[261. Graph Valid Tree](https://leetcode.com/problems/graph-valid-tree/)

- Acyclic
- Connected

## 133. Clone Graph
[133. Clone Graph](https://leetcode.com/problems/clone-graph/)

## 323. Number of Connected Components in an Undirected Graph
[323. Number of Connected Components in an Undirected Graph](https://leetcode.com/problems/number-of-connected-components-in-an-undirected-graph/)

## 684. Redundant Connection
[684. Redundant Connection](https://leetcode.com/problems/redundant-connection/)

## 332. Reconstruct Itinerary
[332. Reconstruct Itinerary](https://leetcode.com/problems/reconstruct-itinerary/)

# Others

## 543. Diameter of Binary Tree
[543. Diameter of Binary Tree](https://leetcode.com/problems/diameter-of-binary-tree/)

## 1245. Tree Diameter
[1245. Tree Diameter](https://leetcode.com/problems/tree-diameter/)

## 1192. Critical Connections in a Network
[1192. Critical Connections in a Network](https://leetcode.com/problems/critical-connections-in-a-network/)

## 399. Evaluate Division
[399. Evaluate Division](https://leetcode.com/problems/evaluate-division/)

## 329. Longest Increasing Path in a Matrix
[329. Longest Increasing Path in a Matrix](https://leetcode.com/problems/longest-increasing-path-in-a-matrix/)

## 2092. Find All People With Secret
[2092. Find All People With Secret](https://leetcode.com/problems/find-all-people-with-secret/)

## 797. All Paths From Source to Target
[797. All Paths From Source to Target](https://leetcode.com/problems/all-paths-from-source-to-target/)

In [14]:
class Solution:
    def allPathsSourceTarget(self, graph):
        """
        Adjacency matrix:
        initial: graph[0]
        DFS
        """
        result = [] 
        if not graph:
            return result 
        
        self.dfs(graph, 0, [], result) 
        return result
    
    def dfs(self, graph, cur_node, curRes, result): 
        # terminate condition 
        if cur_node == len(graph) - 1: 
            curRes.append(cur_node) 
            result.append(copy.deepcopy(curRes)) 
            curRes.pop()
            return 
        
        # find neighbors
        curRes.append(cur_node)
        for nei in graph[cur_node]:
            # print(cur)
            # curRes.append(cur_node)
            self.dfs(graph, nei, curRes, result) 
        curRes.pop()

## Robot Path Planning

In [6]:
visited = [[False for _ in range(2)] for _ in range(3)]
dir = [[0, 1],[1, 0], [0, -1], [-1, 0]]
dir

[[0, 1], [1, 0], [0, -1], [-1, 0]]

In [43]:
from collections import deque
def plan(grid):
    # corner case
    if not grid:
        return ""
    # find the start position
    m = len(grid)
    n = len(grid[0])
    q = deque()
    visited = [[False for _ in range(n)] for _ in range(m)]
    direction = [[0, 1],[0, -1],[1, 0],[-1, 0]]
    d_char = ['E','W','S','N']
    for i in range(m):
        for j in range(n):
            if grid[i][j] == 'S':
                q.append((i, j, ""))
                visited[i][j] = True
                res = bfs(grid, q, visited, direction, d_char)
    return res

def bfs(grid, q, visited, direction, d_char):
    while q:
        cur = q.popleft()
        row = cur[0]
        col = cur[1]
        path = cur[2]
        
        # desitination
        if grid[row][col] == 'G':
            return path
        # build path with it's neighbors
        for i in range(len(direction)):
            r = row + direction[i][0]
            c = col + direction[i][1]
            if r >= 0 and r < len(grid) and c >= 0 and c < len(grid[0]):
                if not visited[r][c] and grid[r][c] != 'O':
                    # generated
                    visited[r][c] = True
                    q.append((r, c, path+d_char[i]))
    return ""
                    
    
a = [['O','E','E','O','E','E'],
     ['O','E','O','O','E','O'],
     ['S','E','E','E','E','E'],
     ['O','E','O','O','E','E'],
     ['O','E','E','E','E','O']]
print(plan(a))

ESSEE


In [40]:
b = [['O','E','E','O','E','E'],
     ['O','E','O','O','E','O'],
     ['S','E','E','E','E','G'],
     ['O','E','O','O','E','E'],
     ['O','E','E','E','E','E']]
m = len(b)
n = len(b[0])
visited = [[False for _ in range(n)] for _ in range(m)]
# print(visited)
# print(m, n)
print(plan(b))